File: repo/backend/app/schemas/governance.py
from datetime import datetime

from pydantic import BaseModel, Field


class DatasetCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    description: str | None = Field(default=None, max_length=5000)
    status: str = Field(default="active", min_length=1, max_length=30)


class DatasetUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    description: str | None = Field(default=None, max_length=5000)
    status: str | None = Field(default=None, min_length=1, max_length=30)


class DatasetOut(BaseModel):
    id: str
    name: str
    description: str | None
    status: str
    created_at: datetime
    updated_at: datetime


class ProjectCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    code: str = Field(min_length=1, max_length=60)
    description: str | None = Field(default=None, max_length=5000)
    status: str = Field(default="active", min_length=1, max_length=30)


class ProjectUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    code: str | None = Field(default=None, min_length=1, max_length=60)
    description: str | None = Field(default=None, max_length=5000)
    status: str | None = Field(default=None, min_length=1, max_length=30)


class ProjectOut(BaseModel):
    id: str
    name: str
    code: str
    description: str | None
    status: str
    created_at: datetime
    updated_at: datetime


class ProjectMemberCreateRequest(BaseModel):
    user_id: str
    role_in_project: str = Field(min_length=1, max_length=60)
    can_edit: bool = True


class ProjectMemberUpdateRequest(BaseModel):
    role_in_project: str | None = Field(default=None, min_length=1, max_length=60)
    can_edit: bool | None = None


class ProjectMemberOut(BaseModel):
    id: str
    user_id: str
    username: str
    role_in_project: str
    can_edit: bool
    created_at: datetime
    updated_at: datetime


class ProjectDatasetOut(BaseModel):
    id: str
    dataset_id: str
    dataset_name: str
    created_at: datetime


class OrgUserOut(BaseModel):
    id: str
    username: str
    active: bool
    roles: list[str]


class AttractionCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    city: str = Field(min_length=1, max_length=120)
    state: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=5000)
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    duration_minutes: int = Field(ge=5, le=720)
    status: str = Field(default="active", min_length=1, max_length=30)


class AttractionUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    city: str | None = Field(default=None, min_length=1, max_length=120)
    state: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=5000)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    duration_minutes: int | None = Field(default=None, ge=5, le=720)
    status: str | None = Field(default=None, min_length=1, max_length=30)


class AttractionOut(BaseModel):
    id: str
    dataset_id: str
    name: str
    city: str
    state: str
    description: str | None
    latitude: float
    longitude: float
    duration_minutes: int
    status: str
    normalized_dedupe_key: str
    merged_into_attraction_id: str | None
    merged_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AttractionDuplicateGroupOut(BaseModel):
    duplicate_key: str
    duplicate_label: str
    candidate_count: int
    candidates: list[AttractionOut]


class AttractionMergeRequest(BaseModel):
    source_attraction_id: str
    target_attraction_id: str
    merge_reason: str | None = Field(default=None, max_length=5000)


class AttractionMergeOut(BaseModel):
    merge_event_id: str
    source_attraction_id: str
    target_attraction_id: str
    merged_at: datetime

File: repo/frontend/src/api/governance.ts
import { apiDelete, apiGet, apiPatch, apiPost } from "./client";

export type Dataset = {
  id: string;
  name: string;
  description: string | null;
  status: string;
  created_at: string;
  updated_at: string;
};

export type Attraction = {
  id: string;
  dataset_id: string;
  name: string;
  city: string;
  state: string;
  description: string | null;
  latitude: number;
  longitude: number;
  duration_minutes: number;
  status: string;
  normalized_dedupe_key: string;
  merged_into_attraction_id: string | null;
  merged_at: string | null;
  created_at: string;
  updated_at: string;
};

export type AttractionDuplicateGroup = {
  duplicate_key: string;
  duplicate_label: string;
  candidate_count: number;
  candidates: Attraction[];
};

export type AttractionMergeResult = {
  merge_event_id: string;
  source_attraction_id: string;
  target_attraction_id: string;
  merged_at: string;
};

export type Project = {
  id: string;
  name: string;
  code: string;
  description: string | null;
  status: string;
  created_at: string;
  updated_at: string;
};

export type OrgUser = {
  id: string;
  username: string;
  active: boolean;
  roles: string[];
};

export type ProjectMember = {
  id: string;
  user_id: string;
  username: string;
  role_in_project: string;
  can_edit: boolean;
  created_at: string;
  updated_at: string;
};

export type ProjectDatasetLink = {
  id: string;
  dataset_id: string;
  dataset_name: string;
  created_at: string;
};

export const governanceApi = {
  listDatasets: () => apiGet<Dataset[]>("/api/datasets"),
  createDataset: (payload: { name: string; description: string; status: string }) =>
    apiPost<Dataset>("/api/datasets", payload),
  updateDataset: (id: string, payload: Partial<{ name: string; description: string; status: string }>) =>
    apiPatch<Dataset>(`/api/datasets/${id}`, payload),

  listProjects: () => apiGet<Project[]>("/api/projects"),
  createProject: (payload: { name: string; code: string; description: string; status: string }) =>
    apiPost<Project>("/api/projects", payload),
  updateProject: (id: string, payload: Partial<{ name: string; code: string; description: string; status: string }>) =>
    apiPatch<Project>(`/api/projects/${id}`, payload),

  listOrgUsers: () => apiGet<OrgUser[]>("/api/admin/users"),

  listAttractions: (datasetId: string) => apiGet<Attraction[]>(`/api/datasets/${datasetId}/attractions`),
  createAttraction: (
    datasetId: string,
    payload: {
      name: string;
      city: string;
      state: string;
      description: string;
      latitude: number;
      longitude: number;
      duration_minutes: number;
      status: string;
    }
  ) => apiPost<Attraction>(`/api/datasets/${datasetId}/attractions`, payload),
  updateAttraction: (
    datasetId: string,
    attractionId: string,
    payload: Partial<{
      name: string;
      city: string;
      state: string;
      description: string;
      latitude: number;
      longitude: number;
      duration_minutes: number;
      status: string;
    }>
  ) => apiPatch<Attraction>(`/api/datasets/${datasetId}/attractions/${attractionId}`, payload),
  listAttractionDuplicates: (datasetId: string) =>
    apiGet<AttractionDuplicateGroup[]>(`/api/datasets/${datasetId}/attractions/duplicates`),
  mergeAttractions: (
    datasetId: string,
    payload: {
      source_attraction_id: string;
      target_attraction_id: string;
      merge_reason?: string;
    }
  ) => apiPost<AttractionMergeResult>(`/api/datasets/${datasetId}/attractions/merge`, payload),

  listProjectMembers: (projectId: string) => apiGet<ProjectMember[]>(`/api/projects/${projectId}/members`),
  addProjectMember: (projectId: string, payload: { user_id: string; role_in_project: string; can_edit: boolean }) =>
    apiPost<ProjectMember>(`/api/projects/${projectId}/members`, payload),
  updateProjectMember: (
    projectId: string,
    memberId: string,
    payload: Partial<{ role_in_project: string; can_edit: boolean }>
  ) => apiPatch<ProjectMember>(`/api/projects/${projectId}/members/${memberId}`, payload),
  removeProjectMember: (projectId: string, memberId: string) =>
    apiDelete<void>(`/api/projects/${projectId}/members/${memberId}`),

  listProjectDatasets: (projectId: string) => apiGet<ProjectDatasetLink[]>(`/api/projects/${projectId}/datasets`),
  linkDatasetToProject: (projectId: string, datasetId: string) =>
    apiPost<ProjectDatasetLink>(`/api/projects/${projectId}/datasets/${datasetId}`, {}),
  unlinkDatasetFromProject: (projectId: string, datasetId: string) =>
    apiDelete<void>(`/api/projects/${projectId}/datasets/${datasetId}`)
};
